src/crystal/tracing.cr

```diff
@@ -199,27 +226,21 @@ module Crystal
       end
     end
 
-    # Formats and prints a log message to stderr. The generated message is
-    # limited to 512 bytes (PIPE_BUF) after which it will be truncated. Being
-    # below PIPE_BUF the message shall be written atomically to stderr,
-    # avoiding interleaved or smashed traces from multiple threads.
-    #
-    # Windows may not have the same guarantees but the buffering should limit
-    # these from happening.
-    macro trace(section, operation, tick = nil, **metadata, &block)
-      if ::Crystal::Tracing.enabled?(\{{section}})
-        %tick = \{{tick}} || ::Crystal::System::Time.ticks
-        \{% if block %}
-          %ret = \{{yield}}
-          %duration = ::Crystal::System::Time.ticks - %tick
-          ::Crystal::Tracing.log(\{{section.id.stringify}}, \{{operation.id.stringify}}, %tick, duration: %duration, \{{metadata.double_splat}})
-          %ret
-        \{% else %}
-          ::Crystal::Tracing.log(\{{section.id.stringify}}, \{{operation.id.stringify}}, %tick, \{{metadata.double_splat}})
-          nil
-        \{% end %}
+    def self.trace(section : Tracing::Section, operation : String, time : UInt64? = nil, **metadata, &)
+      if Tracing.enabled?(section)
+        time ||= System::Time.ticks
+        yield.tap do
```

Comment:
**suggestion:** Consider `ensure` instead of `tap`. In case the block raises, this would still print the tracing log which could be helpful *especially* when something goes wrong. The duration might not be meaningful, but the metadata could be.